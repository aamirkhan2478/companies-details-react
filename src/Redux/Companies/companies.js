import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

// Action Types
const FETCH_COMPANIES = 'FETCH_COMPANIES';
const FETCH_COMPANIES_BY_SYMBOL = 'FETCH_COMPANIES_BY_SYMBOL';

// Reducers
const initialState = {
  isLoading: false,
  companies: [],
  companiesBySymbol: {},
};

const companiesReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case FETCH_COMPANIES:
      return {
        ...state,
        isLoading: false,
        companies: payload,
      };
    case FETCH_COMPANIES_BY_SYMBOL:
      return {
        ...state,
        isLoading: false,
        companiesBySymbol: payload,
      };
    default:
      return state;
  }
};

// Actions
const baseUrl = 'https://financialmodelingprep.com/api/v3';

const fetchCompanies = createAsyncThunk(
  FETCH_COMPANIES,
  async (args, { dispatch }) => {
    try {
      const { data } = await axios.get(
        `${baseUrl}/dowjones_constituent?apikey=c72f83b1dfc59f79c4cdee3c9b93c7f4`,
      );

      const fetchData = data.map((company) => ({
        name: company.name,
        symbol: company.symbol,
        headQuarter: company.headQuarter,
      }));
      dispatch({
        type: FETCH_COMPANIES,
        payload: fetchData,
      });
    } catch (error) {
      console.log(error.message);
    }
  },
);

const fetchBySymbol = createAsyncThunk(
  FETCH_COMPANIES_BY_SYMBOL,
  async (companyName, { dispatch }) => {
    try {
      const { data } = await axios.get(
        `${baseUrl}/profile/${companyName}?apikey=c72f83b1dfc59f79c4cdee3c9b93c7f4`,
      );
      dispatch({
        type: FETCH_COMPANIES_BY_SYMBOL,
        payload: data,
      });
    } catch (error) {
      console.log(error.message);
    }
  },
);

// Exports data
export { fetchCompanies, fetchBySymbol };
export default companiesReducer;
