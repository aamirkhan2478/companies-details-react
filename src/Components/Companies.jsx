import React, { useEffect } from 'react';
import { FaArrowCircleRight } from 'react-icons/fa';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import '../assets/styles/companies.css';
import { v4 as uuidv4 } from 'uuid';
import { fetchCompanies } from '../Redux/Companies/companies';

const Companies = () => {
  const { companies } = useSelector((state) => state.company);
  const navigate = useNavigate();

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCompanies());
  }, []);
  return (
    <>
      <div className="container">
        <h3>Industrial Average</h3>
      </div>
      <div className="bg-pink">
        <h3>Companies Name</h3>
        <div className="row">
          {companies.map((company) => (
            <div
              key={uuidv4()}
              className="col-6"
              onClick={() => navigate(`/details/${company.symbol}`)}
              aria-hidden="true"
            >
              <FaArrowCircleRight className="icon-arrow" />
              <ul className="lists">
                <li>
                  {company.name}
                  <span>
                    (
                    {company.symbol}
                    )
                  </span>
                </li>
                <li>{company.headQuarter}</li>
              </ul>
            </div>
          ))}
        </div>
      </div>
    </>
  );
};

export default Companies;
