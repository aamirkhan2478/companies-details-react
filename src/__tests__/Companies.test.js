import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import store from '../Redux/configureStore';
import Companies from '../Components/Companies';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Companies />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
