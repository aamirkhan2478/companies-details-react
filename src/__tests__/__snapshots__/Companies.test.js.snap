// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="container"
  >
    <h3>
      Industrial Average
    </h3>
  </div>,
  <div
    className="bg-pink"
  >
    <h3>
      Companies Name
    </h3>
    <div
      className="row"
    />
  </div>,
]
`;
